File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Node:
    def __init__(self, node=None):
        self.node = node
        
    def __lt__(self, other):
        if not self.node:
            return other.node
        if not other.node:
            return self.node
        
        return self.node.val < other.node.val

    
class Solution:
    # O(nk), O(k)
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        head, tail = None, None
        
        heap = []
        for node in lists:
            if node:
                heappush(heap, Node(node))
            
        while heap:
            smallest = heappop(heap).node
            
            if smallest.next:
                heappush(heap, Node(smallest.next))
            
            if head:
                tail.next = smallest
                smallest.next = None
                tail = smallest
            else:
                head = smallest
                tail = smallest
            
        return head

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.cpp
/*
keep a heap of all list pointers
pop min and push next until all elements are added to the output list
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class compare {
public:
    bool operator() (ListNode* a, ListNode* b) {
        if (a->val >= b->val) return true;
        else return false;
    }
};

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<ListNode*, vector<ListNode*>, compare> smallest;
        
        ListNode *head = new ListNode();
        ListNode *tail = head;
        
        for (auto& listhead : lists) {
            if (listhead) smallest.push(listhead);
        }
        
        while (!smallest.empty()) {
            ListNode* node = smallest.top();
            smallest.pop();
            
            if (node->next) {
                smallest.push(node->next);
            }
            
            tail->next = node;
            tail = tail->next;
            tail->next = nullptr;
        }
        
        return head->next;
    }
};